### PR TITLE
Add test for MPT tokenization

### PR DESCRIPTION
**Pull request by goerch**
Unfortunately I don't see a way to distinguish `CONTROL` and `USER_DEFINED` tokens when using `AutoTokenizer`.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -7461,6 +7447,21 @@ void llama_sample_repetition_penalties(
...
+        result.resize(-n_tokens);
+        int check = llama_token_to_piece(llama_get_model(ctx), token, result.data(), result.size());
+        GGML_ASSERT(check == -n_tokens);
+    }
+    else {
+        result.resize(n_tokens);
+    }
+
+    return std::string(result.data(), result.size());
+}
```
Comment (ggerganov) [inline llama.cpp:7463]: Any reason to move this down here? Prefer to have it at the top with the rest of the llama helpers

**Review by ggerganov — APPROVED** [review]

File: llama.cpp (hunk 1 above)
Comment (goerch) [reply llama.cpp:7463]: Moved it because `grammar` stuff seems to be the only consumer.

Comment (ggerganov) [reply llama.cpp:7463]: Yes, and we can merge after

File: convert-mpt-hf-to-gguf.py [hunk 2]
```diff
@@ -128,15 +128,21 @@ def parse_args() -> argparse.Namespace:
...
+added_vocab = tokenizer.get_added_vocab()
 reverse_vocab = {id: encoded_tok for encoded_tok, id in tokenizer.vocab.items()}
 
 for i in range(vocab_size):
-    tokens.append(reverse_vocab[i] if i in reverse_vocab else f"[PAD{i}]")
-    scores.append(0.0) # dummy
-    toktypes.append(gguf.TokenType.NORMAL)
+    if i in reverse_vocab:
+        tokens.append(reverse_vocab[i])
+        if reverse_vocab[i] not in added_vocab:
```
Comment (cebtenzzre) [inline convert-mpt-hf-to-gguf.py:137]: It's more straightforward to check if `i` is in `added_vocab.values()`. See my version of the script (written before your BPE changes): https://github.com/nomic-ai/gpt4all/blob/37b007603ab818b596417edcbca0b3dd5e9bc85b/gpt4all-backend/scripts/convert_mpt_hf_to_gguf.py#L101-L103

Comment (goerch) [reply convert-mpt-hf-to-gguf.py]: Tried to clarify the change.

Comment (cebtenzzre) [reply convert-mpt-hf-to-gguf.py]: You're still using `elif reverse_vocab[i] in added_vocab:`, but my version uses `elif i in added_vocab:` and `tokenizer.get_added_vocab().values()`.

Comment (goerch) [reply convert-mpt-hf-to-gguf.py]: Isn't this O(1) vs. O(n)?

**Comment by Galunid** [comment]
I adapted changes from this PR locally on my stablelm branch and they seem to fix the issue from #3604 .

**Comment by goerch** [comment]
> When passing `The best music is ` as a prompt (note ` ` two spaces at the end), they seem to get stripped though
> `The best music isthe music you have the most fun making.`

I think this is what I feared: there are `special` and `non-special` added tokens which have to be handled before the core tokenizer's `encode` and `decode` and we can't distinguish them currently.

File: convert-mpt-hf-to-gguf.py (hunk 2 above)
Comment (cebtenzzre) [reply convert-mpt-hf-to-gguf.py]: You're right, your version is more efficient if there are many added tokens. One could make a set of added token IDs, but I suppose it's simpler to do the dict lookup instead.

**Comment by Galunid** [comment]
I just wanted to let you know, for stablelm use case it's fine, (added_vocab is only spaces and end of text). Thanks for this!

**Comment by goerch** [comment]
> I adapted changes from this PR locally on my stablelm branch and they seem to fix the issue from #3604 .
> 
> When passing `The best music is ` as a prompt (note ` ` two spaces at the end), they seem to get stripped though `The best music isthe music you have the most fun making.`

To fix this issue I imagine to change the detokenizer code to output `USER_DEFINED` tokens (this should repair the problem) and adapt the test cases to ignore `USER_DEFINED` tokens where needed. Will propose this once the test cases are done. Does this sound like a plan?

**Pull request merged**